sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java

```diff
@@ -0,0 +1,239 @@
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.udf;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.crate.exceptions.UserDefinedFunctionAlreadyExistsException;
+import io.crate.exceptions.UserDefinedFunctionUnknownException;
+import io.crate.settings.CrateSetting;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.AckedClusterStateUpdateTask;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateRequest;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static io.crate.operation.udf.UserDefinedFunctionsMetaData.PROTO;
+
+@Singleton
+public class UserDefinedFunctionService {
+
+    static {
+        // register non plugin custom metadata
+        MetaData.registerPrototype(UserDefinedFunctionsMetaData.TYPE, PROTO);
+    }
+
+    public static final CrateSetting<Boolean> UDF_SETTING = CrateSetting.of(Setting.boolSetting(
+        "udf.enabled", false,
+        Setting.Property.NodeScope), DataTypes.BOOLEAN);
+
+    private final Settings settings;
+    private final ClusterService clusterService;
+    private Map<String, UDFLanguage> languageRegistry = new HashMap<>();
+
+    private static final Logger LOGGER = Loggers.getLogger(UserDefinedFunctionService.class);
+
+    @Inject
+    public UserDefinedFunctionService(Settings settings, ClusterService clusterService) {
+        this.settings = settings;
+        this.clusterService = clusterService;
+    }
+
+    public UDFLanguage getLanguage(String languageName) throws IllegalArgumentException {
```

Comment:
I think so. registration happens only on node startup and the language modules would register after another, right. retrieving languages concurrently is not a problem.